python.d/boinc.chart.py

```diff
@@ -0,0 +1,170 @@
+            ['copy_pending', 'Copy Pending', 'absolute', 1, 1]
+        ]
+    }
+}
+
+# A simple template used for pre-loading the return dictionary to make
+# the _get_data() method simpler.
+_DATA_TEMPLATE = {
+    'total': 0,
+    'active': 0,
+    'new': 0,
+    'downloading': 0,
+    'downloaded': 0,
+    'comperror': 0,
+    'uploading': 0,
+    'uploaded': 0,
+    'aborted': 0,
+    'upload_failed': 0,
+    'uninit_sched': 0,
+    'preempted': 0,
+    'scheduled': 0,
+    'uninit_proc': 0,
+    'executing': 0,
+    'suspended': 0,
+    'aborting': 0,
+    'quit': 0,
+    'copy_pending': 0
+}
+
+# Map task states to dimensions
+_TASK_MAP = {
+    boinc_client.ResultState.NEW: 'new',
+    boinc_client.ResultState.FILES_DOWNLOADING: 'downloading',
+    boinc_client.ResultState.FILES_DOWNLOADED: 'downloaded',
+    boinc_client.ResultState.COMPUTE_ERROR: 'comperror',
+    boinc_client.ResultState.FILES_UPLOADING: 'uploading',
+    boinc_client.ResultState.FILES_UPLOADED: 'uploaded',
+    boinc_client.ResultState.ABORTED: 'aborted',
+    boinc_client.ResultState.UPLOAD_FAILED: 'upload_failed'
+}
+
+# Map scheduler states to dimensions
+_SCHED_MAP = {
+    boinc_client.CpuSched.UNINITIALIZED: 'uninit_sched',
+    boinc_client.CpuSched.PREEMPTED: 'preempted',
+    boinc_client.CpuSched.SCHEDULED: 'scheduled',
+}
+
+# Maps process states to dimensions
+_PROC_MAP = {
+    boinc_client.Process.UNINITIALIZED: 'uninit_proc',
+    boinc_client.Process.EXECUTING: 'executing',
+    boinc_client.Process.SUSPENDED: 'suspended',
+    boinc_client.Process.ABORT_PENDING: 'aborted',
+    boinc_client.Process.QUIT_PENDING: 'quit',
+    boinc_client.Process.COPY_PENDING: 'copy_pending'
+}
+
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.order = ORDER
+        self.definitions = CHARTS
+        self.host = self.configuration.get('host', 'localhost')
+        self.port = self.configuration.get('port', 0)
+        self.password = self.configuration.get('password', '')
+        self.client = boinc_client.BoincClient(host=self.host, port=self.port, passwd=self.password)
+        self.alive = False
+
+    def check(self):
+        if platform.system() != 'Linux':
+            self.error('Only supported on Linux.')
+            return False
+        self.connect()
+        self.alive = self.client.connected and self.client.authorized:
+        return self.alive
+
+    def connect(self):
+        self.client.connect()
+
+    def reconnect(self):
+        try:
+            self.client.disconnect()
+        except socket.error:
+            pass
+        self.client.connect()
+        self.alive = self.client.connected and self.client.authorized:
+        return self.alive
+
+    def is_alive(self):
+        if (not self.alive) or \
+           self.client.rpc.sock.getsockopt(socket.IPPROTO_TCP, socket.TCP_INFO, 0) != 1:
```

Comment:
Same as in `spigotmc.chart.py`, this is checking to see if the remote end hung up (more accurately, it's checking if the TCP socket is in the CONNECTED state).

Without either having that there, or the system being configured to use particularly unforgiving TCP shutdown settings, we'll hang when we try to send data over the socket if the remote end randomly disconnects.  Realistically, we _might_ still hang if the network connection is interrupted (as opposed to this which will properly catch the remote end gracefully disconnecting), but we can't reliably compensate for that case because these all get run in threads and therefore we can't setup signal delivery for a timeout.

Reply:
Also, note that this isn't needed in other modules like the beanstalkd one because their interface libraries sanely handle this.  Unfortunately, the `boinc_client.py` code does not, and I don't have a good enough understanding of exactly how it works to update it to do so.

Reply:
Well, add a `TODO` then. This should be fixed i think :smile:

Reply:
Agreed.